File: python/src/binary_market_data/connectors/kalshi_rest.py
"""Public REST polling, deduplicated at the requested output depth."""

import asyncio
from dataclasses import dataclass
from functools import partial
from urllib.parse import quote

import httpx

from .._http import client_context, endpoint, get_text
from ..book import BookError, BookStore, _depth
from ..discovery import KALSHI_REST
from ..types import BookMessage, ConnectionState, ConnectorMessage, StatusMessage, Venue
from ..venue import ParseError, kalshi
from ._common import identifier, managed_stream, positive


@dataclass(frozen=True, slots=True)
class Config:
    market_ticker: str
    rest_base: str = KALSHI_REST
    depth: int = 20
    poll_interval: float = 1.0
    reconnect_max: float = 30.0

    def __post_init__(self) -> None:
        identifier(self.market_ticker)
        endpoint(self.rest_base)
        _depth(self.depth)
        positive(self.poll_interval, "poll_interval")
        positive(self.reconnect_max, "reconnect_max")
        object.__setattr__(self, "poll_interval", max(0.1, self.poll_interval))
        if self.reconnect_max < self.poll_interval:
            raise ValueError("invalid_reconnect_max")

    @classmethod
    def for_market(cls, market_ticker: str) -> "Config":
        return cls(market_ticker)


def stream(config: Config, *, queue_size: int = 256, client: httpx.AsyncClient | None = None):
    """An async context manager yielding messages; exit cancels polling and closes owned IO."""
    return managed_stream(partial(run, config, client=client), queue_size)


async def run(
    config: Config,
    output: asyncio.Queue[ConnectorMessage],
    *,
    client: httpx.AsyncClient | None = None,
) -> None:
    """Advanced API: caller owns task cancellation and must supply a bounded queue."""
    if output.maxsize <= 0:
        raise ValueError("output_queue_must_be_bounded")
    url = endpoint(
        config.rest_base, "markets/" + quote(config.market_ticker, safe="") + "/orderbook"
    )
    connected = False
    last_levels = None
    retry_delay = config.poll_interval
    books = BookStore()
    await output.put(StatusMessage(Venue.KALSHI, ConnectionState.CONNECTING))
    async with client_context(client) as session:
        while True:
            error = None
            try:
                body = await get_text(session, url, params={"depth": str(config.depth)})
                update = kalshi.parse_rest_snapshot(config.market_ticker, body)
                view = books.apply(update, config.depth)
            except ParseError:
                error = "invalid_venue_payload"
            except BookError:
                error = "invalid_book_update"
            except (httpx.HTTPError, ValueError):
                error = "venue_request_failed"
            if error is not None:
                connected = False
                await output.put(
                    StatusMessage(Venue.KALSHI, ConnectionState.DISCONNECTED, detail=error)
                )
                retry_delay = min(retry_delay * 2, config.reconnect_max)
            else:
                if not connected:
                    await output.put(StatusMessage(Venue.KALSHI, ConnectionState.CONNECTED))
                    connected = True
                levels = (view.bids, view.asks)
                if levels != last_levels:
                    last_levels = levels
                    await output.put(BookMessage(view))
                retry_delay = config.poll_interval
            await asyncio.sleep(retry_delay)

File: src/connectors/kalshi_rest.rs
use crate::book::BookStore;
use crate::types::{unix_time_ms, ConnectionState, ConnectorMessage, Venue};
use crate::venue::kalshi;
use std::time::Duration;
use tokio::sync::mpsc;

#[derive(Debug, Clone)]
pub struct Config {
    pub rest_base: String,
    pub market_ticker: String,
    pub depth: usize,
    pub poll_interval: Duration,
    pub reconnect_max: Duration,
}

impl Config {
    pub fn for_market(market_ticker: String) -> Self {
        Self {
            rest_base: "https://external-api.kalshi.com/trade-api/v2".to_string(),
            market_ticker,
            depth: 20,
            poll_interval: Duration::from_secs(1),
            reconnect_max: Duration::from_secs(30),
        }
    }
}

pub async fn run(config: Config, output: mpsc::Sender<ConnectorMessage>) {
    let client = match reqwest::Client::builder()
        .user_agent("binary-market-data/0.1")
        .timeout(Duration::from_secs(10))
        .build()
    {
        Ok(client) => client,
        Err(_error) => {
            send_status(
                &output,
                ConnectionState::Disconnected,
                Some("client_configuration_error".to_string()),
            )
            .await;
            return;
        }
    };
    let encoded_ticker = urlencoding::encode(&config.market_ticker);
    let url = format!(
        "{}/markets/{}/orderbook?depth={}",
        config.rest_base.trim_end_matches('/'),
        encoded_ticker,
        config.depth.min(100)
    );
    let mut books = BookStore::default();
    let mut retry_delay = config.poll_interval;
    let mut connected = false;
    let mut last_levels = None;
    send_status(&output, ConnectionState::Connecting, None).await;

    loop {
        if output.is_closed() {
            return;
        }
        let result = async {
            let response = client.get(&url).send().await?.error_for_status()?;
            response.text().await
        }
        .await;
        match result {
            Ok(body) => match kalshi::parse_rest_snapshot(&config.market_ticker, &body) {
                Ok(update) => match books.apply(update, config.depth) {
                    Ok(view) => {
                        if !connected {
                            send_status(&output, ConnectionState::Connected, None).await;
                            connected = true;
                        }
                        let levels = (view.bids.clone(), view.asks.clone());
                        if last_levels.as_ref() != Some(&levels) {
                            last_levels = Some(levels);
                            if output
                                .send(ConnectorMessage::Book(Box::new(view)))
                                .await
                                .is_err()
                            {
                                return;
                            }
                        }
                        retry_delay = config.poll_interval;
                    }
                    Err(_error) => {
                        connected = false;
                        send_status(
                            &output,
                            ConnectionState::Disconnected,
                            Some("invalid_book_update".to_string()),
                        )
                        .await;
                    }
                },
                Err(_error) => {
                    connected = false;
                    send_status(
                        &output,
                        ConnectionState::Disconnected,
                        Some("invalid_venue_payload".to_string()),
                    )
                    .await;
                }
            },
            Err(_error) => {
                connected = false;
                send_status(
                    &output,
                    ConnectionState::Disconnected,
                    Some("venue_request_failed".to_string()),
                )
                .await;
                retry_delay = retry_delay.saturating_mul(2).min(config.reconnect_max);
            }
        }
        tokio::select! {
            _ = tokio::time::sleep(retry_delay) => {}
            _ = output.closed() => return,
        }
    }
}

async fn send_status(
    output: &mpsc::Sender<ConnectorMessage>,
    state: ConnectionState,
    detail: Option<String>,
) {
    let _ = output
        .send(ConnectorMessage::Status {
            venue: Venue::Kalshi,
            state,
            at_ms: unix_time_ms(),
            detail,
        })
        .await;
}
